bridgenode/rev.go

```diff
@@ -0,0 +1,348 @@
+// RevBlock is the structure of how a block is stored in the
+// rev*.dat file the Bitcoin Core generates
+type RevBlock struct {
+	Magic [4]byte   // Network magic bytes
+	Size  [4]byte   // size of the BlockUndo record
+	Txs   []*TxUndo // acutal undo record
+	Hash  [32]byte  // 32 byte double sha256 hash of the block
+}
+
+// TxUndo contains the TxInUndo records.
+// see github.com/bitcoin/bitcoin/src/undo.h
+type TxUndo struct {
+	TxIn []*TxInUndo
+}
+
+// TxInUndo is the stucture of the undo transaction
+// Eveything is uncompressed here
+// see github.com/bitcoin/bitcoin/src/undo.h
+type TxInUndo struct {
+	Height int32
+
+	// Version of the original tx that created this tx
+	Varint uint64
+
+	// scriptPubKey of the spent UTXO
+	PKScript []byte
+
+	// Value of the spent UTXO
+	Amount int64
+
+	// Whether if the TxInUndo is a coinbase or not
+	// Not actually included in the rev*.dat files
+	Coinbase bool
+}
+
+// Deserialize takes a reader and reads a single block
+// Only initializes the Block var in RevBlock
+func (rb *RevBlock) Deserialize(r io.Reader) error {
+	txCount, err := wire.ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+
+	for i := uint64(0); i < txCount; i++ {
+		var tx TxUndo
+		err := tx.Deserialize(r)
+		if err != nil {
+			return err
+		}
+		rb.Txs = append(rb.Txs, &tx)
+	}
+	return nil
+}
+
+// Deserialize takes a reader and reads all the TxUndo data
+func (tx *TxUndo) Deserialize(r io.Reader) error {
+
+	// Read the Variable Integer
+	count, err := wire.ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+
+	for i := uint64(0); i < count; i++ {
+		var in TxInUndo
+		err := readTxInUndo(r, &in)
+		if err != nil {
+			return err
+		}
+		tx.TxIn = append(tx.TxIn, &in)
+	}
+	return nil
+}
+
+// readTxInUndo reads all the TxInUndo from the reader to the passed in txInUndo
+// variable
+func readTxInUndo(r io.Reader, ti *TxInUndo) error {
+	// nCode is how height is saved to the rev files
+	nCode, _ := deserializeVLQ(r)
+	ti.Height = int32(nCode / 2) // Height is saved as actual height * 2
+	ti.Coinbase = nCode&1 == 1   // Coinbase is odd. Saved as height * 2 + 1
+
+	// Only TxInUndos that have the height greater than 0
+	// Has varint that isn't 0. see
+	// github.com/bitcoin/bitcoin/blob/9cc7eba1b5651195c05473004c00021fe3856f30/src/undo.h#L42
+	if ti.Height > 0 {
+		varint, err := wire.ReadVarInt(r, pver)
+		if err != nil {
+			return err
+		}
+		if varint != 0 {
+			return fmt.Errorf("varint is %d", varint)
+		}
+		ti.Varint = varint
+	}
+
+	amount, _ := deserializeVLQ(r)
+	ti.Amount = decompressTxOutAmount(amount)
+
+	ti.PKScript = decompressScript(r)
+	if ti.PKScript == nil {
+		panic("nil pkscript")
+	}
+
+	return nil
+}
+
+// OpenIndexFile returns the db with only read only option enabled
+func OpenIndexFile() *leveldb.DB {
+	s := "/home/calvin/.bitcoin/testnet3/blocks/index"
+	lvdb, err := leveldb.OpenFile(s, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	ro := new(opt.ReadOptions)
+	ro.DontFillCache = true
```

Comment:
Is the db really opened with readonly? https://github.com/syndtr/goleveldb/blob/758128399b1df3a87e92df6c26c1d2063da8fabe/leveldb/opt/options.go#L609-L614

also seems like `ro` isn't used

Reply:
Ah yeah something to clean up before un-drafting. Thanks!

Reply:
addressed